### PR TITLE
FIX: Don't revert the entirety of `sys.modules` after running `kernprof.main`

### LINKED ISSUE #436: Reverting `sys.modules` causes pickling issues with `concurrent.futures.ProcessPoolWorker`
## Summary

When working on #431 I ran into a weird situation where the in-process integration tests using `concurrent.futures` would occasionally fail when one picks a random subset thereof to run, and pass if the entire test module is run, indicating some global-state pollution. After some spelunking it seems that the cause is:

- `kernprof` imports `concurrent.futures`, which (< Python 3.15) uses a module-level `__getattr__()` to lazy-load the various `Executor` subclasses (e.g. `ProcessPoolExecutor`) when requested.
- In `kernprof.py::_main_profile()` and likewise `line_profiler/autoprofile/autoprofile.py::run()`, we essentially force-revert the state of `sys.modules` to before these functions are called.
- If `concurrent.futures.process` hasn't already been loaded before `kernprof.py::main()` is called, but the profiled code imports `ProcessPoolExecutor` from `concurrent.futures`:
  - `concurrent.futures.process` will be loaded inside `kernprof.py::main()`, and its copy of `ProcessPoolExecutor` will be cached in the `concurrent.futures` namespace.
  - When `kernprof.py::main()` exits, `concurrent.futures.process` will be dropped from `sys.modules`, but its `ProcessPoolExecutor` remains in `concurrent.futures`.
- If one is to go on and make a second identical call to `kernprof.py::main()`:
  - The code will load the copy of `ProcessPoolExecutor` belonging to the now-dead copy of `concurrent.futures.process` which (1) its methods all refer to and (2) the import system doesn't know about.
  - `ProcessPoolExecutor` will then create processes with `Process(target=concurrent.futures.process._process_worker, ...)`, which then promptly goes on to fail with a `PicklingError` because its copy of `_process_worker()` is stale and would not be the same object as  `sys.modules['concurrent.futures.process']._process_worker`.

The freezing of `sys.modules` was originally (#339, #423) intended to mirror the functionalities of `runpy.run_module(alter_sys=True)`, but it seems that we've been overzealous: should've stuck to just reverting `sys.modules['__main__']`, and tempering with the rest may after all cause more problems than it solves. While this bug may seem specific, a package lazy-loading from its submodules via `__getattr__()` is a common pattern, and it's not hard to imagine other such cases where import orders causes unexpected behaviors.

Will write a small PR fix.

## Example

<details>
<summary><tt>demo.py</tt></summary>

```python
import os
from tempfile import TemporaryDirectory

from kernprof import main as kp_main


script = """
from __future__ import annotations

from multiprocessing import get_context
from concurrent.futures import ProcessPoolExecutor


def my_sum(x: list[int]) -> int:
    result = 0
    for n in x:
        result += n
    return result


def main() -> None:
    with ProcessPoolExecutor(mp_context=get_context('spawn')) as ex:
        print(list(ex.map(my_sum, [[1, 2], [3, 4], [5, 6]])))


if __name__ == '__main__':
    main()
"""


def main(n: int = 2) -> None:
    for _ in range(n):
        with TemporaryDirectory() as tmpdir:
            kp_main([
                '-lzv',
                f'--outfile={os.path.join(tmpdir, "out.lprof")}',
                '-c',
                script,
            ])


if __name__ == '__main__':
    main()
```

</details>

<details>
<summary> Result </summary>

```shell
$ python demo.py
[3, 7, 11]
Wrote profile results to '[...]/out.lprof'
Timer unit: ...
...
Function: main at line 15
...
Wrote profile results to '[...]/out.lprof'
Timer unit: ...
...
Function: main at line 15
...
Traceback (most recent call last):
  ...
_pickle.PicklingError: Can't pickle <function _process_worker at ...>: it's not the same object as concurrent.futures.process._process_worker
when serializing dict item '_target'
when serializing multiprocessing.context.SpawnProcess state
when serializing multiprocessing.context.SpawnProcess object
...
```

</details>

### PR DESCRIPTION
**Pull request by TTsangSC**
Closes #436.

(**EDIT**: typos; missing summaries for `kernprof.y::main()` and `line_profiler/autoprofile/autoprofile.py::run()`)

## Code changes

- `line_profiler/line_profiler_utils.py::restore`:
  - Migrated from `kernprof.py::_restore`, so that other parts of the codebase can use it for cleanup.
  - The constructor method `.mapping()` (resp. `.instance_dict()`) now takes an extra argument `keys: Collection[K]` (resp. `attrs: Collection[str]`) to limit the scope of reversion.
  - The context manager is now reentrant.
  - Minor backward incompatibility: instances no longer expose the attributes `.getter`, `.setter`, and `.old`. (But we don't really use it anywhere other than as anonymous decorators and context managers anyway, so nothing changed.)
- `kernprof.py::main()`:
  On exit:
  - Instead of freezing the entire `sys.modules`, we now only revert `sys.modules['__main__']` or the lack thereof.
  - Instead of freezing the entire `line_profiler._diagnostics` namespace, we now only revert `line_profiler._diagnostics.log` which is overridden.
 - `line_profiler/autoprofile/autoprofile.py::run()`:
   Instead of freezing the entire `sys.modules` when running the supplied code, we now only revert `sys.modules['__main__']` or the lack thereof.

## Test changes

- `line_profiler/line_profiler_utils.py::restore.sequence()`'s doctest extended to demonstrate/check for reentrant behavior.
- `tests/test_kernprof.py`:
  - Factored out common pattern of creating a tempdir and `cd`-ing therein into a helper context manager (`chdir_temp`).
  - Minor fix: `test_kernprof_m_import_resolution()` and `test_kernprof_eager_preimport_bad_module()` now use `os.pathsep` instead of a hard-coded `':'` to join `${PYTHONPATH}` components.
  - New test `test_ppe_pickling` formalizes the failing case in #436, checking that repeated invocation of `kernprof.main()` running  `concurrent.futures` code doesn't fail due to consistency issues with `sys.modules`.

### CODE REVIEW

File: tests/test_kernprof.py [hunk 1]
```diff
@@ -542,6 +557,83 @@ def test_call_with_diagnostics(module, builtin):
...
+
+    def my_sum(x: list[int]) -> int:
+        result = 0
+        for n in x:
+            result += n
+        return result
+
+
+    def main() -> None:
+        with ProcessPoolExecutor(mp_context=get_context('spawn')) as ex:
```
Comment (Erotemic) [inline tests/test_kernprof.py]:
```suggestion
        with ProcessPoolExecutor(max_workers=2, mp_context=get_context('spawn')) as ex:
```

Not sure if we want to use all CPUs on a test. On one hand it could expose more edge cases on the other, setting it is more stable.

File: kernprof.py [hunk 2]
```diff
@@ -228,6 +228,9 @@ def main():
     positive_float,
     short_string_path,
 )
+from line_profiler.line_profiler_utils import (
+    restore as _restore,  # Compatibility
```
Comment (Erotemic) [inline kernprof.py]: Compatibility for a private name isn't required. It doesn't bother me though. This is mostly a nitpick.

File: tests/test_kernprof.py [hunk 3]
```diff
@@ -188,14 +221,13 @@ def main():
...
-        temp_dpath = ub.Path(tmpdir)
+    with chdir_temp() as temp_dpath:
         namespace_mod_path = temp_dpath / 'my_namesapce_pkg' / 'mysubmod.py'
         namespace_mod_path.parent.mkdir()
         namespace_mod_path.write_text(code)
-        python_path = tmpdir
+        python_path = str(temp_dpath)
         if 'PYTHONPATH' in os.environ:
-            python_path += ':' + os.environ['PYTHONPATH']
+            python_path += os.pathsep + os.environ['PYTHONPATH']
```
Comment (Erotemic) [inline tests/test_kernprof.py:230]: good catch

**Review by Erotemic — APPROVED** [review]
This all LGTM. Minor comments. You can choose what to do with them. Say the word and I'll merge.

File: kernprof.py (hunk 2 above)
Comment (TTsangSC) [reply kernprof.py]: Fair enough, since we're technically already changing `restore`'s API (it no longer has `.getter`, `.setter`, and `.old` – but then again those members aren't used anywhere by the code outside of the class itself). Might as well just import under the new name.

**Comment by TTsangSC** [comment]
Thanks for the quick review, should be alright now (other than the `ty` lint – but that's a known issue, and #434 would patch `pyproject.toml` and fix that).

**Comment by Erotemic** [comment]
Yeah I ignored the ty error for that reason, let's land #434 next
